### DISCUSSION TITLE
Should `gcc` be required on Fedora Silverblue?

### DISCUSSION THREAD

**Discussion by osalbahr** [Homebrew/discussions#5627] [Linux] [answered]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.23-53-g29c22e0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 29c22e0ab3fd9826cefd4b14c143266dea3bc391
Last commit: 2 days ago
Core tap JSON: 22 Sep 12:14 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.4 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin/ruby
CPU: octa-core 64-bit sandybridge
Clang: N/A
Git: 2.46.1 => /bin/git
Curl: 8.6.0 => /bin/curl
Kernel: Linux 6.10.10-200.fc40.x86_64 x86_64 GNU/Linux
OS: Fedora release 40 (Forty)
Host glibc: 2.39
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: 14.2.0
xorg: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No developer tools installed.
Install Clang or run `brew install gcc`.
```

### Description of issue

I am unsure why `gcc` is required when installing `neofetch` on Fedora Silverblue ([`gcc` is not a dependency for `neofetch`](https://formulae.brew.sh/formula/neofetch)).
```console
$ brew deps neofetch
binutils
gcc
gmp
isl
libmpc
lz4
mpfr
xz
zlib
zstd
```

I was able to find in [the docs](https://docs.brew.sh/Homebrew-on-Linux) that:

> If you’re using an older distribution of Linux, installing your first package will also install a recent version of glibc and gcc. Use brew doctor to troubleshoot common issues.

However I don't know why Fedora Silverblue would be considered too old. Although `glibc` wasn't installed by Homebrew, so Homebrew correctly identified only that "gcc is too old" (by virtue of not existing). Development on Fedora Silverblue is [intended happen by using containers](https://docs.fedoraproject.org/en-US/fedora-silverblue/toolbox) but "overlaying" gcc is also the option. I am happy to do the latter if Homebrew does need `gcc`, but I would rather not do that if Homebrew (or installed formulae) don't actually need it.

> Homebrew does not use any libraries provided by your host system, except glibc and gcc if they are new enough. Homebrew can install its own current versions of glibc and gcc for older distributions of Linux.

But why? Why might `neofetch` need `gcc`? It is a compiler, not a runtime dependency. Is this a legacy behavior, intended for older systems, or intended for all systems?

Also, I think Fedora Silverblue does satisfy the "Requirements" section. So I expected only "actual" dependencies to be installed.

Let me know if there is something I am missing.

**Comment by Bo98** [comment, marked as answer]
Short answer is that it should probably be broken down better but it currently doesn't because (a) nobody has worked on it and (b) we don't have separate subpackages anyway - just a single `gcc` formula.

Homebrew could depend on any of the following depending on the formula (with the whatever corresponding minimum versions roughly align with GCC 11):

* libgcc_s
* libstdc++
* libgomp
* libquadmath
* GCC compiler itself (e.g. if you installed ruby and ran `gem install` on a native gem)

(+ any glibc library or anything from `lib6c-dev`/`glibc-devel` or equivalent)

Because the compiler itself usually covers all of it, we checked for that for simplicity. I agree that this should be broken down better - it just isn't yet and that would require some work to do. There is currently no metadata attached to formulae that is able to deduce which of the above is needed ahead-of-time.

**Reply by osalbahr** [reply]
Yeah. This also seems to be one of those "would be nice" but there's a reason it hasn't been done yet. It's often a combination of high difficulty and low importance. Thx for clarifying!

For those curious: I will likely go the `rpm-ostree install gcc` route.
- My intuition says this is less storage-heavy than Homebrew's `gcc` since Homebrew (rightfully so) makes very few assumptions about the system, but I haven't done the math (see https://github.com/Homebrew/brew/issues/18373).
